crates/accelerate/src/basis/basis_translator/compose_transforms.rs

```diff
@@ -0,0 +1,206 @@
+use qiskit_circuit::circuit_instruction::OperationFromPython;
+use qiskit_circuit::converters::circuit_to_dag;
+use qiskit_circuit::imports::{GATE, PARAMETER_VECTOR, QUANTUM_CIRCUIT, QUANTUM_REGISTER};
+use qiskit_circuit::operations::OperationRef;
+use qiskit_circuit::packed_instruction::PackedInstruction;
+use qiskit_circuit::parameter_table::ParameterUuid;
+use qiskit_circuit::Qubit;
+use qiskit_circuit::{
+    circuit_data::CircuitData,
+    dag_circuit::{DAGCircuit, NodeType},
+    operations::{Operation, Param},
+};
+use smallvec::SmallVec;
+
+// Custom types
+// TODO: Remove these and use the version from `basis_search`
+pub type BasisTransforms = Vec<(String, u32, SmallVec<[Param; 3]>, CircuitRep)>;
+pub type MappedTransforms = HashMap<(String, u32), (SmallVec<[Param; 3]>, DAGCircuit)>;
+// TODO: Remove these and use the version from `EquivalenceLibrary`
+
+/// Representation of QuantumCircuit which the original circuit object + an
+/// instance of `CircuitData`.
+#[derive(Debug, Clone)]
+pub struct CircuitRep(pub CircuitData);
+
+impl FromPyObject<'_> for CircuitRep {
+    fn extract_bound(ob: &Bound<'_, PyAny>) -> PyResult<Self> {
+        if ob.is_instance(QUANTUM_CIRCUIT.get_bound(ob.py()))? {
+            let data: Bound<PyAny> = ob.getattr("_data")?;
+            let data_downcast: Bound<CircuitData> = data.downcast_into()?;
+            let data_extract: CircuitData = data_downcast.extract()?;
+            Ok(Self(data_extract))
+        } else {
+            Err(PyTypeError::new_err(
+                "Provided object was not an instance of QuantumCircuit",
+            ))
+        }
+    }
+}
+
+impl IntoPy<PyObject> for CircuitRep {
+    fn into_py(self, py: Python<'_>) -> PyObject {
+        QUANTUM_CIRCUIT
+            .get_bound(py)
+            .call_method1("_from_circuit_data", (self.0,))
+            .unwrap()
+            .unbind()
+    }
+}
+
+impl ToPyObject for CircuitRep {
+    fn to_object(&self, py: Python<'_>) -> PyObject {
+        self.clone().into_py(py)
+    }
+}
+
+#[pyfunction(name = "compose_transforms")]
+pub(super) fn py_compose_transforms(
+    py: Python,
+    basis_transforms: BasisTransforms,
+    source_basis: HashSet<(String, u32)>,
+    source_dag: &DAGCircuit,
+) -> PyResult<MappedTransforms> {
+    compose_transforms(py, &basis_transforms, &source_basis, source_dag)
+}
+
+pub(super) fn compose_transforms<'a>(
+    py: Python,
+    basis_transforms: &'a BasisTransforms,
+    source_basis: &'a HashSet<(String, u32)>,
+    source_dag: &'a DAGCircuit,
+) -> PyResult<MappedTransforms> {
+    let example_gates = *get_example_gates(py, source_dag, None)?;
+    let mut mapped_instructions: MappedTransforms = HashMap::new();
+
+    for (gate_name, gate_num_qubits) in source_basis.iter().cloned() {
+        // Need to grab a gate instance to find num_qubits and num_params.
+        // Can be removed following https://github.com/Qiskit/qiskit-terra/pull/3947 .
```

Comment:
Yes, this was a question I had. Not sure how to tackle it since it seems that issue has been abandoned.

Reply:
I would probably just remove the comment, and add a method comment to `get_example_gates` that explains that its purpose is just to get the number of parameters for a gate with a specific name and number of qubits, since that info is only found on an instruction.

(...also, we should probably rename `get_example_gates` to `get_num_params` and change the return type to  `HashMap<(String, u32), usize>`. As it is, we don't use the `PackedInstruction`s coming out of the function (do we?) and they are being cloned)